File: src/test/java/no/nav/sosialhjelp/soknad/business/service/adressesok/AdresseStringSplitterTest.java
package no.nav.sosialhjelp.soknad.business.service.adressesok;

import no.nav.sosialhjelp.soknad.client.kodeverk.KodeverkService;
import org.junit.jupiter.api.Test;

import java.util.Objects;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class AdresseStringSplitterTest {

    @Test
    void tomStrengGirBlanktSvar() {
        assertThat(AdresseStringSplitter.toSokedata(null,"").adresse).isBlank();
    }
    
    @Test
    void nullStrengGirNullSvar() {
        assertThat(AdresseStringSplitter.toSokedata(null, null).adresse).isNull();
    }
    
    @Test
    void kunAdresseVirker() {
        assertThat(AdresseStringSplitter.toSokedata(null, "asdf").adresse).isEqualTo("asdf");
    }
    
    @Test
    void husnummer() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "asdf 2");
        assertThat(result.adresse).isEqualTo("asdf");
        assertThat(result.husnummer).isEqualTo("2");
    }

    @Test
    void kunHusnummer() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "234");
        assertThat(result.husnummer).isEqualTo("234");
    }
    
    @Test
    void husbokstav() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "asdf 2G");
        assertThat(result.adresse).isEqualTo("asdf");
        assertThat(result.husnummer).isEqualTo("2");
        assertThat(result.husbokstav).isEqualTo("G");
    }

    @Test
    void husnummerOgBokstav() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "212G");
        assertThat(result.husnummer).isEqualTo("212");
        assertThat(result.husbokstav).isEqualTo("G");
    }
    
    @Test
    void postnummer() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "asdf 2G, 0882");
        assertThat(result.adresse).isEqualTo("asdf");
        assertThat(result.husnummer).isEqualTo("2");
        assertThat(result.husbokstav).isEqualTo("G");
        assertThat(result.postnummer).isEqualTo("0882");
    }

    @Test
    void postnummerMedMellomromFlyttet() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "asdf 2G ,0882");
        assertThat(result.adresse).isEqualTo("asdf");
        assertThat(result.husnummer).isEqualTo("2");
        assertThat(result.husbokstav).isEqualTo("G");
        assertThat(result.postnummer).isEqualTo("0882");
    }

    @Test
    void kunPostnummer() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "0882");
        assertThat(result.postnummer).isEqualTo("0882");
    }

    @Test
    void kunPostnummerMedMellomrom() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "   0882   ");
        assertThat(result.postnummer).isEqualTo("0882");
    }
    
    @Test
    void poststed() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "asdf 2G, 0882 OSLO");
        assertThat(result.adresse).isEqualTo("asdf");
        assertThat(result.husnummer).isEqualTo("2");
        assertThat(result.husbokstav).isEqualTo("G");
        assertThat(result.postnummer).isEqualTo("0882");
        assertThat(result.poststed).isEqualTo("OSLO");
    }

    @Test
    void kunGateOgPostnummer() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "Veivei, 0110 ");
        assertThat(result.adresse).isEqualTo("Veivei");
        assertThat(result.postnummer).isEqualTo("0110");
    }

    @Test
    void kunGateOgPoststed() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "Veivei, OSLO");
        assertThat(result.adresse).isEqualTo("Veivei");
        assertThat(result.poststed).isEqualTo("OSLO");
    }

    @Test
    void dobbeltnavnPlussDiverseMellomrom() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "    Nedre Glommas    Vei   211G  ,  0882  ØVRE OSLO   ");
        assertThat(result.adresse).isEqualTo("Nedre Glommas Vei");
        assertThat(result.husnummer).isEqualTo("211");
        assertThat(result.husbokstav).isEqualTo("G");
        assertThat(result.postnummer).isEqualTo("0882");
        assertThat(result.poststed).isEqualTo("ØVRE OSLO");
    }
    @Test
    void kompakt() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, " Nedre Glommas Vei211G,0882ØVRE OSLO   ");
        assertThat(result.adresse).isEqualTo("Nedre Glommas Vei");
        assertThat(result.husnummer).isEqualTo("211");
        assertThat(result.husbokstav).isEqualTo("G");
        assertThat(result.postnummer).isEqualTo("0882");
        assertThat(result.poststed).isEqualTo("ØVRE OSLO");
    }

    @Test
    void poststedUtenPostnummer() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "asdf 2G OSLO");
        assertThat(result.adresse).isEqualTo("asdf");
        assertThat(result.husnummer).isEqualTo("2");
        assertThat(result.husbokstav).isEqualTo("G");
        assertThat(result.poststed).isEqualTo("OSLO");
    }
    
    @Test
    void utenHusnummer() {
        final Sokedata result = AdresseStringSplitter.toSokedata(null, "asdf, 0882 OSLO");
        assertThat(result.adresse).isEqualTo("asdf");
        assertThat(result.postnummer).isEqualTo("0882");
        assertThat(result.poststed).isEqualTo("OSLO");
    }
    
    @Test
    void skalKunneSokeMedKommunenavn() {
        final KodeverkService kodeverkService = mock(KodeverkService.class);
        when(kodeverkService.gjettKommunenummer(anyString())).thenReturn("0301");
        final Sokedata result = AdresseStringSplitter.toSokedata(kodeverkService, "asdf, OSLO");
        assertThat(result.adresse).isEqualTo("asdf");
        assertThat(result.poststed).isNull();
        assertThat(result.kommunenummer).isEqualTo("0301");
    }
    
    @Test
    void skalFungereMedPoststedSelvMedKodeverk() {
        final KodeverkService kodeverkService = mock(KodeverkService.class);
        when(kodeverkService.gjettKommunenummer(anyString())).thenReturn("0301");
        final Sokedata result = AdresseStringSplitter.toSokedata(kodeverkService, "asdf, 0756 OSLO");
        assertThat(result.adresse).isEqualTo("asdf");
        assertThat(result.postnummer).isEqualTo("0756");
        assertThat(result.poststed).isEqualTo("OSLO");
        assertThat(result.kommunenummer).isNull();
    }

    @Test
    void postnummerMatchTest() {
        Sokedata sokedata = AdresseStringSplitter.postnummerMatch("0001");
        assertThat(Objects.requireNonNull(sokedata).postnummer).isEqualTo("0001");
        sokedata = AdresseStringSplitter.postnummerMatch("0001 ");
        assertThat(Objects.requireNonNull(sokedata).postnummer).isEqualTo("0001");
        sokedata = AdresseStringSplitter.postnummerMatch(" 0001");
        assertThat(Objects.requireNonNull(sokedata).postnummer).isEqualTo("0001");

        sokedata = AdresseStringSplitter.postnummerMatch("Haugeveien, 0001 klavestaad");
        assertThat(sokedata).isNull();
        sokedata = AdresseStringSplitter.postnummerMatch("Sannergata 2");
        assertThat(sokedata).isNull();
        sokedata = AdresseStringSplitter.postnummerMatch("Sannergata0001");
        assertThat(sokedata).isNull();
        sokedata = AdresseStringSplitter.postnummerMatch("0001Klavestad");
        assertThat(sokedata).isNull();
        sokedata = AdresseStringSplitter.postnummerMatch("0001 Klavestad");
        assertThat(sokedata).isNull();
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/adressesok/sok/AdresseStringSplitterTest.kt
package no.nav.sosialhjelp.soknad.adressesok.sok

import io.mockk.every
import io.mockk.mockk
import no.nav.sosialhjelp.soknad.adressesok.sok.AdresseStringSplitter.toSokedata
import no.nav.sosialhjelp.soknad.kodeverk.KodeverkService
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

internal class AdresseStringSplitterTest {

    @Test
    fun tomStrengGirBlanktSvar() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "")
        assertThat(sokedata!!.adresse).isBlank
    }

    @Test
    fun nullStrengGirNullSvar() {
        val sokedata = toSokedata(kodeverkService = null, adresse = null)
        assertThat(sokedata!!.adresse).isNull()
    }

    @Test
    fun kunAdresseVirker() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "asdf")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
    }

    @Test
    fun husnummer() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
        assertThat(sokedata.husnummer).isEqualTo("2")
    }

    @Test
    fun kunHusnummer() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "234")
        assertThat(sokedata!!.husnummer).isEqualTo("234")
    }

    @Test
    fun husbokstav() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2G")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
        assertThat(sokedata.husnummer).isEqualTo("2")
        assertThat(sokedata.husbokstav).isEqualTo("G")
    }

    @Test
    fun husnummerOgBokstav() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "212G")
        assertThat(sokedata!!.husnummer).isEqualTo("212")
        assertThat(sokedata.husbokstav).isEqualTo("G")
    }

    @Test
    fun postnummer() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2G, 0882")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
        assertThat(sokedata.husnummer).isEqualTo("2")
        assertThat(sokedata.husbokstav).isEqualTo("G")
        assertThat(sokedata.postnummer).isEqualTo("0882")
    }

    @Test
    fun postnummerMedMellomromFlyttet() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2G ,0882")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
        assertThat(sokedata.husnummer).isEqualTo("2")
        assertThat(sokedata.husbokstav).isEqualTo("G")
        assertThat(sokedata.postnummer).isEqualTo("0882")
    }

    @Test
    fun poststed() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2G, 0882 OSLO")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
        assertThat(sokedata.husnummer).isEqualTo("2")
        assertThat(sokedata.husbokstav).isEqualTo("G")
        assertThat(sokedata.postnummer).isEqualTo("0882")
        assertThat(sokedata.poststed).isEqualTo("OSLO")
    }

    @Test
    fun kunGateOgPostnummer() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "Veivei, 0110 ")
        assertThat(sokedata!!.adresse).isEqualTo("Veivei")
        assertThat(sokedata.postnummer).isEqualTo("0110")
    }

    @Test
    fun kunGateOgPoststed() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "Veivei, OSLO")
        assertThat(sokedata!!.adresse).isEqualTo("Veivei")
        assertThat(sokedata.poststed).isEqualTo("OSLO")
    }

    @Test
    fun dobbeltnavnPlussDiverseMellomrom() {
        val sokedata = toSokedata(
            kodeverkService = null,
            adresse = "    Nedre Glommas    Vei   211G  ,  0882  ØVRE OSLO   "
        )
        assertThat(sokedata!!.adresse).isEqualTo("Nedre Glommas Vei")
        assertThat(sokedata.husnummer).isEqualTo("211")
        assertThat(sokedata.husbokstav).isEqualTo("G")
        assertThat(sokedata.postnummer).isEqualTo("0882")
        assertThat(sokedata.poststed).isEqualTo("ØVRE OSLO")
    }

    @Test
    fun kompakt() {
        val sokedata = toSokedata(kodeverkService = null, adresse = " Nedre Glommas Vei211G,0882ØVRE OSLO   ")
        assertThat(sokedata!!.adresse).isEqualTo("Nedre Glommas Vei")
        assertThat(sokedata.husnummer).isEqualTo("211")
        assertThat(sokedata.husbokstav).isEqualTo("G")
        assertThat(sokedata.postnummer).isEqualTo("0882")
        assertThat(sokedata.poststed).isEqualTo("ØVRE OSLO")
    }

    @Test
    fun poststedUtenPostnummer() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2G OSLO")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
        assertThat(sokedata.husnummer).isEqualTo("2")
        assertThat(sokedata.husbokstav).isEqualTo("G")
        assertThat(sokedata.poststed).isEqualTo("OSLO")
    }

    @Test
    fun utenHusnummer() {
        val sokedata = toSokedata(kodeverkService = null, adresse = "asdf, 0882 OSLO")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
        assertThat(sokedata.postnummer).isEqualTo("0882")
        assertThat(sokedata.poststed).isEqualTo("OSLO")
    }

    @Test
    fun skalKunneSokeMedKommunenavn() {
        val kodeverkService = mockk<KodeverkService>()
        every { kodeverkService.gjettKommunenummer(any()) } returns "0301"

        val sokedata = toSokedata(kodeverkService = kodeverkService, adresse = "asdf, OSLO")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
        assertThat(sokedata.poststed).isNull()
        assertThat(sokedata.kommunenummer).isEqualTo("0301")
    }

    @Test
    fun skalFungereMedPoststedSelvMedKodeverk() {
        val kodeverkService = mockk<KodeverkService>()
        every { kodeverkService.gjettKommunenummer(any()) } returns "0301"

        val sokedata = toSokedata(kodeverkService = kodeverkService, adresse = "asdf, 0756 OSLO")
        assertThat(sokedata!!.adresse).isEqualTo("asdf")
        assertThat(sokedata.postnummer).isEqualTo("0756")
        assertThat(sokedata.poststed).isEqualTo("OSLO")
        assertThat(sokedata.kommunenummer).isNull()
    }

    @Test
    fun `postnummer fra 1 til 4 tegn`() {
        var sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2G, 0")
        assertThat(sokedata!!.postnummer).isEqualTo("0")

        sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2G, 03")
        assertThat(sokedata!!.postnummer).isEqualTo("03")

        sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2G, 030")
        assertThat(sokedata!!.postnummer).isEqualTo("030")

        sokedata = toSokedata(kodeverkService = null, adresse = "asdf 2G, 0301")
        assertThat(sokedata!!.postnummer).isEqualTo("0301")
    }
}
